Add CRL cache module for OpenSSL-style CRL directories

legoscia opened the pull request:
This pull request replaces #968. (It is based on `master` instead of `maint`, and [you can't change the target branch of a pull request](https://github.com/isaacs/github/issues/18).)

Compared to the previous version, I've made the `ssl_crl_cache_api` change backwards compatible: if the callback module doesn't have `lookup/3`, `lookup/2` is used as before; and made the code more readable using bit syntax. I also thought about the names of the hash functions. While OpenSSL has the only other implementation that I know of (the Oracle tool `orapki` also generates hash file names for CRLs, but I don't know what algorithm it uses), there is nothing stopping this from being used in other situations. The best name I could think of is "short hash", since that's what it is: eight hexadecimal digits. Thus we have `short_crl_issuer_hash` and `short_cert_issuer_hash`, implemented in terms of `short_name_hash`. What do you think? Any better ideas?

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
We think the names are better :)  However we still think there are some improvements to be done. 

We do not like that you call internal funtions in asn-1 asn1rt_nif:decode_ber_tlv. You should handle it in the way done in pubkey_cert_records.erl 

Bitsyntax supports Endianness,  <<X:32/little>> (big is default)

Why not use integer_to_list(X, 16).   instead of lists:flatten(io_lib:format("~8.16.0b", [HashValue])). 
maybe with a pad function to be  equivalent with io_lib-variant.

To make the code "upgrade safe"  you need to do a try M:F(A1, A2, A3 ) ...  catch error:undef  -> M:F(A1, A2).  Look at ssl_manager.erl  do_register_session/6 for example.

legoscia commented:
Completely forgot about the endianness for bit syntax :) I also addressed the other points. I managed to make the code produce the slightly invalid output required for the hash (UTF-8 strings everywhere) in more or less the same style as in pubkey_cert_records.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
Thank you, those parts look good now.  :) So now we have some new points and questions/considerations for you.

* The function short_hash_name, is it interesting to call this function as a user of public_key? You say in the docs that it is used by the other two functions and then we think it needs not be an exported function if that is its use case. It is easy to add API functions later if needed, but tedious to remove API functions.

* ssl_crl_hash_dir is an alternate callback module to the default one and that supports it own option,
but the module does not have any API functions, and should not have its own manual page. You need to put the description in the ssl manual page  under the crl_cache option description.

* The function match_dp uses lots of non API functions in public_key. As this code is located in the ssl application this is a problem. On the other hand public_key has a lot of pkix_* functions to handle
X509-structures. So maybe something along the lines of pkix_match_dist_point?!

* The function normalise_name is perhaps something that  should be merged with the function
public_key:pkix_normalize_name/1 ?

Keep up the good work.

legoscia commented:
Thanks for the feedback! I redid the documentation for `ssl_crl_hash_dir`, and added a function `public_key:pkix_match_dist_point`, to avoid internal `public_key` calls in `ssl_crl_hash_dir`.

As for the API of `public_key`, `short_name_hash` is really the only function that `ssl_crl_hash_dir` needs, since it gets passed the issuer name, not any certificate or CRL. I added the other functions partly because I needed them to implement the test cases, and partly because they would make it possible to implement something like `c_rehash` in Erlang, should that ever be desired, but that's not actually needed in the use case I have. Thus, I added an extra commit that removes those two functions and leaves `short_name_hash`. What do you think? If this seems to be a good idea, I'll squash it into the earlier commit.

I thought about combining this with `public_key:pkix_normalize_name`, but I'm leaning towards this being a bad idea. The main reason is that `normalise_name` produces names that don't follow the specification and thus can't be encoded (UTF-8 strings everywhere). While it seems like this wouldn't cause any problems in the current code base, it's imaginable that a program would take the return value of `public_key:pkix_dist_point` (which calls `pkix_normalize_name`) and expect to be able to include that in a certificate to be encoded, which would be broken by this change.

I'm leaving for Easter holidays now; will be back late next week.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
We think that it is a good if this PR adds only what is need for its use case. If someone would like to implement something like c_rehash and sees a need to export more functions as API functions a new PR can made that does that, so squash it. 

Ok I see your point about normalize name, it was too long ago that I wrote the normalize code for me to realize this. You have a valid point, I agree. Could you perhaps think of a name for your normalize name function that reflects this or at least write a comment in the code that make people aware of this.

Happy Easter

legoscia commented:
There, removed the superfluous functions, and rearranged the internal ones a bit and added some comments. Hopefully that makes it clearer what they are for, and how they are used.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
We will have another look soon, we have been a bit preoccupied. But our aim is that it will make it to 19.0

IngelaAndin commented:
We decided to move this PR into the test stage.

IngelaAndin commented:
Due to many changes in this area we have not been able to test this yet, stalled in wait of gen_statem rewrite.

IngelaAndin commented:
Looking fairly good in the builds now, one problem remains.

*** CT Error Notification 2016-05-17 08:18:32.527 ***
make_certs:eval_cmd failed on line 335
Reason: {eval_cmd,"openssl ca -gencrl  -crlhours -24 -out /ldisk/daily_build/19_master-opu_o.2016-05-16_20/test/test_server/ct_run.test_server@frerin.2016-05-17_08.05.03/test.ssl_test.logs/run.2016-05-17_08.05.10/log_private/crl_hash_dir/CRL-maybe-expired-CA/crl.pem -config /ldisk/daily_build/19_master-opu_o.2016-05-16_20/test/test_server/ct_run.test_server@frerin.2016-05-17_08.05.03/test.ssl_test.logs/run.2016-05-17_08.05.10/log_private/crl_hash_dir/CRL-maybe-expired-CA/ca.cnf",
          1}

This happens on an openbsd machine with  OpenSSL version: LibreSSL 2.3.2  
I have not had time to investigate further, but it happened before that we had to make an OpenSSL configfile more "secure", we removed a environment variable if I remember correctly, to make LibreSSL happy.

IngelaAndin commented:
My coworker has found that the option "-crlhours -24"  is wrong. 
There should not be a  "-"  before "24"

Could you please fix that. We are now aiming to get better statistics on github and if I fix it the automatic close will not work and the statistics looks less good.

legoscia commented:
Hm, the `-24` thing is a little hack necessary for the test: it is supposed to create a CRL that expired 24 hours in the past. Apparently LibreSSL doesn't like that.

Not sure if there's a way to get LibreSSL to create expired CRLs. Perhaps the best solution is to skip those tests in that case...

legoscia commented:
Like this, perhaps? The test gets skipped for LibreSSL, but still passes with OpenSSL.

IngelaAndin commented:
In the long run I think the right thing to do is add functionality to public_key so that we can sign
CRLS  with help of public_key. For cert expired test in ssl  I update the cert validity in Erlang and use public_key:pkix_sign to make a new cert that has expired. But your solution may be acceptable for now, as we do not have time for additional features for 19.0.  and also such an addition can be considered an other contribution outside the scoop of this one. Additional it is good to have test

IngelaAndin commented:
Or maybe it is possible to use  -enddate option  see https://www.openssl.org/docs/manmaster/apps/ca.html

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

legoscia commented:
I tried using the `-enddate` option instead of `-crlhours`, but both OpenSSL and LibreSSL fail, saying `cannot lookup how long until the next CRL is issued`. It seems like `-enddate` only applies when signing certificates, not when creating CRLs, for which [only `-crldays`, `-crlhours` and `-crlsec` are used](https://github.com/libressl-portable/openbsd/blob/49f0814b1bf2a75568cdc59cc6a2601e885ff0b7/src/usr.bin/openssl/ca.c#L1193-L1196).

IngelaAndin commented:
Now we get other errors :( 

With the OpenSSL versions  OpenSSL 1.0.0j-fips 10 May 2012   and OpenSSL 1.0.1p-freebsd 9 Jul 2015   the test case crl_hash_dir_collision fails  with the following error log

```
*** System report during ssl_crl_SUITE:crl_hash_dir_collision/1 in crl_hash_dir 2016-05-31 04:18:29.337 ***

=ERROR REPORT==== 31-May-2016::06:18:29 ===
SSL: certify: ssl_handshake.erl:420:Fatal error: certificate unknown - failed_to_decode_certificate


*** User 2016-05-31 04:18:29.338 ***
Server Alert or Close {error,{tls_alert,"certificate unknown"}}


*** System report during ssl_crl_SUITE:crl_hash_dir_collision/1 in crl_hash_dir 2016-05-31 04:18:29.338 ***

=ERROR REPORT==== 31-May-2016::06:18:29 ===
SSL: certify: ssl_alert.erl:97:Fatal error: certificate unknown


*** CT Error Notification 2016-05-31 04:18:29.339 ***
ssl_crl_SUITE:crl_hash_dir_collision failed on line 339
Reason: {test_case_failed,{{expected,{<0.10237.0>,
                              {error,{tls_alert,"certificate revoked"}}}},
                   {got,{<0.10237.0>,
                         {error,{tls_alert,"certificate unknown"}}}}}}
```

legoscia commented:
Hm, that's a strange error. I built OpenSSL 1.0.0j (couldn't find a fips version) locally, and couldn't reproduce it. Is it possible that this is an intermittent error, that only appears for those two versions by chance, as the tests are run on a wide of systems?

By the line number in the error message, this is failing on the third test with the same certificates, just the CRLs have changed. Would it be possible to rerun the tests with the following patch (obviously not for inclusion)? Or failing that, would it be possible to get the contents of the `log_private/crl_hash_dir/crls` directory? One of those files should be a CRL that fails to be parsed.

```
--- a/lib/ssl/src/ssl_handshake.erl.~upstream_master~	2016-06-02 15:28:02.000000000 +0100
+++ b/lib/ssl/src/ssl_handshake.erl	2016-06-02 18:17:53.000000000 +0100
@@ -415,8 +415,10 @@
 		path_validation_alert(Reason)
 	end
     catch
-	error:_ ->
+	error:Error ->
 	    %% ASN-1 decode of certificate somehow failed
+	    error_logger:error_report([{failed_to_decode_certificate, Error},
+				       {stacktrace, erlang:get_stacktrace()}]),
             ?ALERT_REC(?FATAL, ?CERTIFICATE_UNKNOWN, failed_to_decode_certificate)
     end.
 

```

IngelaAndin commented:
Yes it is a strange error, we have been struggling to reach zero failing test cases on all platforms. And seen a lot strange errors with some windows compilers and certain openssl version that just disappear when we use a mys-compiled version. I also see some other strange unkonow_CA errors in other places that might be related. I will make a debug branch with your suggested  debugg printout and see if it might help both our problems.

IngelaAndin commented:
I will send you the contents of privdir as a mail.

Error message is:
```

=ERROR REPORT==== 7-Jun-2016::08:26:21 ===
    failed_to_decode_certificate: {badmatch,
                                   {error,
                                    {asn1,
                                     {badmatch,{error,issuer_not_found}}}}}
    stacktrace: [{public_key,der_encode,2,
                             [{file,"public_key.erl"},{line,263}]},
                 {ssl_certificate,certificate_chain,3,
                                  [{file,"ssl_certificate.erl"},{line,99}]},
                 {ssl_crl,trusted_cert_and_path,3,
                          [{file,"ssl_crl.erl"},{line,44}]},
                 {pubkey_crl,verify_crl,9,
                             [{file,"pubkey_crl.erl"},{line,149}]},
                 {pubkey_crl,validate,7,[{file,"pubkey_crl.erl"},{line,35}]},
                 {public_key,do_pkix_crls_validate,5,
                             [{file,"public_key.erl"},{line,983}]},
                 {ssl_handshake,crl_check_same_issuer,4,
                                [{file,"ssl_handshake.erl"},{line,2123}]},
                 {ssl_handshake,'-validation_fun_and_state/6-fun-1-',7,
                                [{file,"ssl_handshake.erl"},{line,1472}]}]

```

legoscia commented:
I found the immediate reason for that crash. If the function `ssl_crl:find_issuer` would return `{error, issuer_not_found}`, then the second clause of `ssl_crl:trusted_cert_and_path` would treat that return value as a certificate (since it's returned, not thrown), and pass it on to `ssl_certificate:certificate_chain`, which would try to encode it, failing with `asn1 badmatch`.

I still don't understand why this would only happen sometimes... I'll investigate further and see what I find.

legoscia commented:
So that crash was actually pointing to some kind of underlying error. `pubkey_crl:verify_crl` was trying to verify that the CRL was signed, and in the course of doing so asked for the certificate for the `hash-collision-0000000000` CA, which for whatever reason was not in the ETS table, even though it was sent by the server in the certificate message. Something strange is going on...

IngelaAndin commented:
I have done some modifications to ssl_crl.erl so that it shall not hide the error. Currently I also think it is strange that this only happens sometimes... we will keep investigating. It would be nice it this on could make it for 19. It so close now!

IngelaAndin commented:
Looking much better now, I suspect  propagating the error correctly back to the path_validation 
made the difference. Now there is one little problem left, on a few machines this happens

`*** System report during ssl_crl_SUITE:crl_hash_dir_collision/1 in crl_hash_dir 2016-06-09 06:51:00.769 ***

=INFO REPORT==== 9-Jun-2016::08:51:00 ===
    application: ssl
    exited: stopped
    type: temporary




=== Ended at 2016-06-09 06:51:00
=== Location: [{ssl_crl_SUITE,crl_verify_valid,421},
              {ssl_crl_SUITE,crl_hash_dir_collision,319},
              {test_server,ts_tc,1529},
              {test_server,run_test_case_eval1,1045},
              {test_server,run_test_case_eval,977}]
=== Reason: {failed,
                  {test_case_failed,
                   {{expected,{<0.11277.0>,ok}},
                    {expected,
                     {{connect_failed,{tls_alert,"certificate expired"}},ok}},
                    {got,
                     {<0.11277.0>,
                      {badrpc,
                       {'EXIT',
                        {function_clause,
                         [{ssl,send,
                           [{error,closed},"Hello world"],
                           [{file,"ssl.erl"},{line,276}]},
                          {ssl_test_lib,send_recv_result_active,1,
                           [{file,"ssl_test_lib.erl"},{line,1098}]},
                          {rpc,local_call,3,[{file,"rpc.erl"},{line,311}]},
                          {ssl_test_lib,do_run_server,3,
                           [{file,"ssl_test_lib.erl"},{line,93}]}]}}}}}}}}

`
and 
`*** System report during ssl_crl_SUITE:crl_hash_dir_expired/1 in crl_hash_dir 2016-06-09 06:51:38.894 ***

=INFO REPORT==== 9-Jun-2016::08:51:38 ===
    application: ssl
    exited: stopped
    type: temporary




=== Ended at 2016-06-09 06:51:38
=== Location: [{ssl_crl_SUITE,crl_verify_valid,421},
              {ssl_crl_SUITE,crl_hash_dir_expired,392},
              {test_server,ts_tc,1529},
              {test_server,run_test_case_eval1,1045},
              {test_server,run_test_case_eval,977}]
=== Reason: {failed,
                  {test_case_failed,
                   {{expected,{<0.11741.0>,ok}},
                    {expected,
                     {{connect_failed,{tls_alert,"certificate expired"}},ok}},
                    {got,
                     {<0.11741.0>,
                      {badrpc,
                       {'EXIT',
                        {function_clause,
                         [{ssl,send,
                           [{error,closed},"Hello world"],
                           [{file,"ssl.erl"},{line,276}]},
                          {ssl_test_lib,send_recv_result_active,1,
                           [{file,"ssl_test_lib.erl"},{line,1098}]},
                          {rpc,local_call,3,[{file,"rpc.erl"},{line,311}]},
                          {ssl_test_lib,do_run_server,3,
                           [{file,"ssl_test_lib.erl"},{line,93}]}]}}}}}}}}

`

legoscia commented:
I haven't managed to reproduce that error, but I was able to reproduce the opposite problem, and I don't understand why this happens.

Using the files in the tarball you sent me, and removing `crl_hash_dir/crls/1627b4b0.r1` which seems to be from a different test case, I run this little server in one Erlang shell:

```erlang
ssl:start().

f(ListenSocket).
Who = "/tmp/log_private/crl_hash_dir/CRL-maybe-expired",
{ok, ListenSocket} =
ssl:listen(0,
           [{keyfile, Who ++ "/key.pem"},
            {certfile, Who ++ "/cert.pem"},
            {cacertfile, Who ++ "/cacerts.pem"}]).
io:format("Listening on ~p~n", [ssl:sockname(ListenSocket)]).


F = fun F (LS) -> case ssl:transport_accept(LS, 10000000) of {ok, NewSocket} -> ssl:ssl_accept(NewSocket), F(LS) end end.
F(ListenSocket).
```

And in another Erlang shell I run this client code, adjusting the port argument:

```erlang
(fun(Port) ->
         ssl:stop(), ssl:start(),
         Options = [{cacertfile, "/tmp/log_private/crl_hash_dir/CRL-maybe-expired-CA/cacerts.pem"},
                    {verify, verify_peer},
                    {crl_cache, {ssl_crl_hash_dir, {internal, [{dir, "/tmp/log_private/crl_hash_dir/crls"}]}}}],
         {ok, S1} = ssl:connect({127,0,0,1}, Port, [{crl_check, best_effort}]++Options),
         io:format("success 1\n"),
         {error,{tls_alert,"bad certificate"}} = ssl:connect({127,0,0,1}, Port, [{crl_check, true}]++Options),
         io:format("failure 2\n")
end)(12345).
```

Since the only available CRL is expired, I expect the `best_effort` connection to succeed and the `true` connection to fail, but in fact both connections succeed. If I remove the `best_effort` connection attempt, then the `true` connection attempt fails as expected. Tracing reveals that no functions in the `ssl_crl_hash_dir` module are called during the second connection attempt, suggesting that the positive CRL check result is somehow cached, but I can't see where that could possibly happen...

The errors you posted are the other way around: a `best_effort` connection fails when the only available CRL is expired, when we'd expect it to ignore that CRL.

I'll keep investigating next week.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

IngelaAndin commented:
We will keep investigating and making the tests work everywhere and would also appreciate possible further involvement from your side, however we decided this will not be
a show stopper for this contribution to make it to 19. We do not think this causes any new problems that can effect users that not uses the new functionality.

IngelaAndin commented:
Humm ... seems you have rebased the branch. Have you changed anything that ought to be included?
It seems I merged an older version of the branch?

legoscia commented:
Oh, I rebased it to get rid of the merge conflict. I just removed the last commit, the one fixing `ssl_crl:find_issuer`, as it conflicted with the master branch. What's in the master branch now looks good. (Except for a missing comma in the export list in `public_key_SUITE`.)

Thanks a lot for your work with this! I'll try to track down those last failing tests.

IngelaAndin commented:
Comma should be fixed now, that was embarrassing!

IngelaAndin commented:
I will close the PR and will keep in touch

legoscia commented:
I had another look at this, and noticed something strange. In the test output, there are two different timestamps: Common Test reports 6 AM, but the info report says 8 AM - presumably UTC and CEST, respectively. However, the certificate itself has a validity period that starts at 8 AM GMT - so it is reported as "certificate expired" because the certificate isn't valid _yet_. My best guess is that some combination of OpenSSL version and weird time settings in the OS caused this: OpenSSL thinks that the local time is UTC, but Erlang doesn't....

rtraschke commented:
I think you can set the info log times to be UTC by setting the `sasl` env var `utc_log` to `true`.

IngelaAndin commented:
Humm ...  this might have been on our time-warp test machine, in which case this is not an issue. There has been so many strange failures in the builds lately, but it looks to be clearing up now and it seems most problems have been test environment problems. Thank you very much for your efforts, I think you can stop looking now. I let you know if there are any problems that persist when we cuddled all the test machines, but I think it looks promising.
